zkevm-circuits/src/evm_circuit/execution/error_oog_sload_sstore.rs

```diff
@@ -0,0 +1,585 @@
+            // value_prev != value, original_value == value, original_value != 0
+            let testing_data = TestingData::new_for_sstore(
+                TESTING_STORAGE_KEY,
+                0x060504.into(),
+                0x060505.into(),
+                0x060504.into(),
+                is_warm,
+            );
+            test_root(&testing_data);
+            test_internal(0x20, 0x00, &testing_data);
+            test_internal(0x1010, 0xff, &testing_data);
+        });
+    }
+
+    #[test]
+    fn test_oog_sstore_reset_inexistent() {
+        [false, true].into_iter().for_each(|is_warm| {
+            // value_prev != value, original_value == value, original_value == 0
+            let testing_data = TestingData::new_for_sstore(
+                TESTING_STORAGE_KEY,
+                0.into(),
+                0x060505.into(),
+                0.into(),
+                is_warm,
+            );
+            test_root(&testing_data);
+            test_internal(0x20, 0x00, &testing_data);
+            test_internal(0x1010, 0xff, &testing_data);
+        });
+    }
+
+    #[test]
+    fn test_oog_sstore_recreate_slot() {
+        [false, true].into_iter().for_each(|is_warm| {
+            // value_prev != value, original_value != value_prev, original_value != value,
+            // value_prev == 0
+            let testing_data = TestingData::new_for_sstore(
+                TESTING_STORAGE_KEY,
+                0x060504.into(),
+                0x0.into(),
+                0x060506.into(),
+                is_warm,
+            );
+            test_root(&testing_data);
+            test_internal(0x20, 0x00, &testing_data);
+            test_internal(0x1010, 0xff, &testing_data);
+        });
+    }
+
+    #[test]
+    fn test_oog_sstore_recreate_slot_and_reset_inexistent() {
+        [false, true].into_iter().for_each(|is_warm| {
+            // value_prev != value, original_value != value_prev, original_value == value,
+            // value_prev == 0
+            let testing_data = TestingData::new_for_sstore(
+                TESTING_STORAGE_KEY,
+                0x060504.into(),
+                0x0.into(),
+                0x060504.into(),
+                is_warm,
+            );
+            test_root(&testing_data);
+            test_internal(0x20, 0x00, &testing_data);
+            test_internal(0x1010, 0xff, &testing_data);
+        });
+    }
+
+    #[derive(Default)]
+    struct TestingData {
+        key: U256,
+        value: U256,
+        value_prev: U256,
+        original_value: U256,
+        is_warm: bool,
+        gas_cost: u64,
+        bytecode: Bytecode,
+    }
+
+    impl TestingData {
+        pub fn new_for_sload(key: U256, is_warm: bool) -> Self {
+            let mut bytecode = bytecode! {
+                PUSH32(key)
+                SLOAD
+            };
+            let mut gas_cost =
+                OpcodeId::PUSH32.constant_gas_cost().0 + cal_sload_gas_cost_for_assignment(false);
```

Comment:
this line should use param  `is_warm` rather than hard code `false`?
`OpcodeId::PUSH32.constant_gas_cost().0 + cal_sload_gas_cost_for_assignment(is_warm);`

Reply:
Seems no. The first `sload(key)` is `is_warm == false` (cold key). If the parameter `is_warm == true` (warm key), then the gas cost of second `sload(key)` is `cal_sload_gas_cost_for_assignment(true)`.

The bytecode is set for `is_warm == true` as:
```
PUSH32(key)
SLOAD  // cal_sload_gas_cost_for_assignment(false)
PUSH32(key)
SLOAD  // cal_sload_gas_cost_for_assignment(true)
```

Reply:
I see, with a little confusion.